File: Elixir/problem009.exs
# Problem 9 - Project Euler
# http://projecteuler.net/index.php?section=problems&id=9
defmodule P009 do
  def triplet?(x, y) do
    z2 = x * x + y * y
    z = trunc(:math.sqrt(z2))
    if z * z == z2  do
      z
    else
      nil
    end
  end

  def solve() do
    3..1000
    |> Stream.map(
         fn x ->
           x + 1..1000
           |> Stream.filter(&(&1 * &1 + x * x <= 1000000))
           |> Stream.map(
                fn y ->
                  z = triplet?(x, y)
                  if z && (x + y + z == 1000) do
                    x * y * z
                  end
                end
              )
           |> Stream.filter(&(&1))
           |> Enum.at(0)
         end
       )
    |> Stream.filter(&(&1))
    |> Enum.at(0)
  end
end

IO.puts P009.solve()

File: Go/problem009.go
package projecteuler

import "fmt"

// #9 Special Pythagorean triplet - Project Euler
// http://projecteuler.net/index.php?section=problems&id=9

func P009() {
	for a := 1; a < 1000; a++ {
		for b := a; b < 1000; b++ {
			c := 1000 - a - b
			if c <= b {
				break
			}
			if a*a+b*b == c*c {
				fmt.Println(a * b * c)
				return
			}
		}
	}
}
